src/orange/OrangeParams.cc

```diff
@@ -180,15 +178,63 @@ OrangeParams::OrangeParams(OrangeInput input)
             }
             volume_labels.push_back(std::move(volume_label));
         }
-
-        bbox_ = u.bbox;
     }
 
     surf_labels_ = LabelIdMultiMap<SurfaceId>{std::move(surface_labels)};
     vol_labels_ = LabelIdMultiMap<VolumeId>{std::move(volume_labels)};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Advanced usage: construct from explicit host data.
+ *
+ * Volume and surface labels must be unique for the time being.
+ */
+OrangeParams::OrangeParams(OrangeInput input)
+{
+    CELER_VALIDATE(!input.units.empty(), << "input geometry has no universes");
+
+    HostVal<OrangeParamsData> host_data;
 
-    supports_safety_ = host_data.simple_unit[SimpleUnitId{0}].simple_safety;
-    bbox_ = input.units.front().bbox;
+    // Copy data from OrangeInput
+    host_data.scalars.max_level = input.max_level;
+    make_builder(&host_data.universe_types)
+        .insert_back(input.universe_types.begin(), input.universe_types.end());
+    make_builder(&host_data.universe_indices)
+        .insert_back(input.universe_indices.begin(),
+                     input.universe_indices.end());
+
+    // Insert simple units
+    insert_simple_units(host_data, input);
+
+    // Insert rect array Units
+    detail::RectArrayInserter insert_rect_array(&host_data);
+    for (RectArrayInput const& r : input.rect_arrays)
+    {
+        CELER_VALIDATE(
+            r, << "rect array '" << r.label << "' is not properly constructed");
+        insert_rect_array(r);
+    }
+
+    switch (input.universe_types[0])
+    {
+        case UniverseType::simple: {
```

Comment:
I don't think this is an inconvenient limitation, but why does a RectArray not have a well-defined exterior?

Reply:
You'd need a special treatment for everything inside the bounds versus outside: all the interior cells are convex; the outside is not. It also means dedicating another cell index to the exterior which would be only used in the special case of an exterior rect array.

Reply:
Yeah so it could be supported (it doesn't seem that difficult), but moreso there isn't a use case. If there was a use case, making an additional SimpleUnit surrounding the array would't be burdensome.